docs/ldml/tr35-keyboards.md

```diff
@@ -1314,11 +1322,18 @@ _Attribute:_ `form` (required)
 > This attribute specifies the physical layout of a hardware keyboard,
 > or that the form is a `touch` layout.
 >
-> It is recommended to always have at least one hardware (non-touch) form.
+> When using an on-screen keyboard, if there is not a `<layers form="touch">`
+> element, the hardware elements can be used for on-screen use. However, a `touch` keyboard
+> cannot be used with a hardware keyboard.
```

Comment:
There are fancy hardware keyboards that have software controlled keycaps

Reply:
> There are fancy hardware keyboards that have software controlled keycaps

But generally they are still not touch keyboards with arbitrary layout -- they have a fixed hardware layout, and they generate known scan codes by position, no different to German vs French labelling on a traditional ISO hardware keyboard, where the electronics are identical, and it's just the text printed on the key caps that are changing.

Reply:
"cannot be used" is not clear what it means, does it mean "implementations must reject" or "not guaranteed to work" or "undefined behavior"?

Reply:
I would expect this to be "implementations must reject"

Reply:
If the hardware keyboard is really a big touch panel, then it's more of a touch keyboard.

Reply:
"cannot be used" and "reject" are both ill-defined (here and elsewhere). Is it non-conformant to use a touch keyboard layout with a hardware keyboard? Or is it just a ill-advised, because it cannot reasonably be interpreted in a way that would be useful with a hardware keyboard? 

And what would it mean for an implementation to "reject" a keyboard. I know what it would mean to have an application for a mortgage to be "rejected". But a software data structure? I could understand "must not use", but that also seems overloaded. If a touch keyboard layout simply couldn't apply to hardware keyboard, that's like saying you must not use a skateboard on a ski-slope.

Reply:
> what would it mean for an implementation to "reject" a keyboard.

My interpretation of that is user tries to load a xml file into OS and OS says no, cannot use this file.

Reply:
If you mean that, you should say it. But is that a hard requirement on an OS, that it "must say 'no, you cannot use "this file"'. And does that mean an application (non-OS) that supports keyboards can use it?

Reply:
That is what I was asking in the first place...

My preference would be we leave that to whoever is trying to load the file. Perhaps they decide they can deal with it, perahps they decide it's safer to not attempt.

Reply:
I think it's a little stronger than that. We deliberately designed the touch layouts to be completely independent of hardware layouts -- from scan codes on up. Attempting to jam a touch layout into a hardware layout through some algorithmic process may kinda work, for some layouts. But I want to avoid the expectation that this will even be possible -- it just doesn't really make sense and we shouldn't be leaving that interpretation open on the data model. And if we do leave it open, we may end up with implementers pushing back on more novel touch layout designs because it breaks their touch-to-hardware algorithms.

That said, I'd be happy for this to be expressed as a implementation recommendation or guideline if that's preferable to others.

Reply:
please see what you think of the current recommendation